### PR TITLE
doc: update `toPythonApplication` section according to RFC 140

### PR DESCRIPTION
**Pull request by AndersonTorres**
The advent of `pkgs/by-name` hierarchy created an interesting alternative: instead of cluttering `all-packages.nix` with more lines and more merge conflicts, the creation of `toPythonApplication` expressions can be made at `pkgs/by-name`.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by AndersonTorres** [comment]
Marking
- @infinisil as mastermind of by-name hierarchy
- @ShamrockLee as beta-tester at https://github.com/NixOS/nixpkgs/pull/291176

**Comment by lolbinarycat** [comment]
calling it an alternative doesn't seem right, isn't `by-name` the preferred way of doing things now?

**Comment by ShamrockLee** [comment]
This format provides a more robust way to override dependent python packages, and also make the dependent `toPythonPackage` overridable.

However, we do hope to maintain a consistent overriding interface. Considering the following scenario:
Let's say that we have `pkgs.youtube-dl` packaged in this format:

```nix
{ python3Packages }:
with python3Packages; toPythonApplication youtube-dl
```

where `python3Packages` is `python311Packages`. If, for some reason, `python3Packages.youtube-dl` breaks, and the maintainers decide to switch to `python310Packages.youtube-dl`, should we add a line in `all-packages.nix` and specify `youtube-dl = callPackage pkgs/by-name/yo/youtube-dl/package.nix { python3Packages = python310Packages; }`, or should we just change the argument from `python3Packages` to `python310Packages`?

**Comment by lolbinarycat** [comment]
i don't actually know python, so i'm not much help here, but if such a fundamental issue as "how do we package for different python versions" isn't solved yet, then it sounds like this packaging method isn't ready.

**Comment by ShamrockLee** [comment]
BTW, why do we prefer `python3*Packages` over `python3*.pkgs` while we have to override it by overriding `python3*` and then access `python3*.pkgs`? After the adoption of this pattern, we'll need to do such kind of overriding often, and it's necessary to have a sensible way of overriding.

**Comment by ShamrockLee** [comment]
> i don't actually know python, so i'm not much help here, but if such a fundamental issue as "how do we package for different python versions" isn't solved yet, then it sounds like this packaging method isn't ready.

I personally prefer the second approach, i.e. pin the version in `all-packages.nix` and remove the line to un-pin it. My point is that when it comes to the choice between keeping overriding interface consistent and avoiding `all-packages.nix`, the overriding interface consistency should win.

**Comment by infinisil** [comment]
I think using `toPythonApplication` like this is great! The only thing to watch out for is whether this causes a change in the `.override` interface when you switch from a `toPythonApplication` in `all-packages.nix` to one in `package.nix`. I believe it does, so to prevent breakages you'd have to only use this for new packages or so.

> where `python3Packages` is `python311Packages`. If, for some reason, `python3Packages.youtube-dl` breaks, and the maintainers decide to switch to `python310Packages.youtube-dl`, should we add a line in `all-packages.nix` and specify `youtube-dl = callPackage pkgs/by-name/yo/youtube-dl/package.nix { python3Packages = python310Packages; }`, or should we just change the argument from `python3Packages` to `python310Packages`?

This is indeed a concern, and I agree with your conclusion:

> I personally prefer the second approach, i.e. pin the version in `all-packages.nix` and remove the line to un-pin it. My point is that when it comes to the choice between keeping overriding interface consistent and avoiding `all-packages.nix`, the overriding interface consistency wins.

This is also what is recommended in [`pkgs/by-name/README.md`](https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#changing-implicit-attribute-defaults).

**Comment by AndersonTorres** [comment]
> calling it an alternative doesn't seem right, isn't `by-name` the preferred way of doing things now?

`toPythonApplication` cals are usually inlined at `all-packages.nix`.

**Comment by AndersonTorres** [comment]
Ops, my failure!

**Comment by AndersonTorres** [comment]
Ops, my failure!

**Comment by infinisil** [comment]
(Discussed in the docs team meeting today)

> > calling it an alternative doesn't seem right, isn't `by-name` the preferred way of doing things now?
> 
> `toPythonApplication` cals are currently inlined at `all-packages.nix`.

I think it's fine for something to be recommended even if older instances don't use the pattern (yet).

File: doc/languages-frameworks/python.section.md [hunk 1]
```diff
@@ -344,6 +344,16 @@ applied to the reference:
 youtube-dl = with python3Packages; toPythonApplication youtube-dl;
 ```
 
+Since the advent of `pkgs/by-name` hierarchy, there is an extra alternative:
+instead of writing the code above in `all-packages.nix`, the corresponding code
```
Comment (DanielSidhion) [inline doc/languages-frameworks/python.section.md:348]: Given the agreement that this is the preferred way to declare these types of packages going forward, can you reword this to fit in the previous paragraph (instead of "A reference shall be created...", reword to fit this text there instead)?

And then we can ellaborate on:
- Existing packages and why they must remain in `all-packages.nix`. I'd go with a note first (`:::{.note}`), but if the text becomes too large, can be a subsection.
- What to do in cases we need to override the package

Comment (AndersonTorres) [reply doc/languages-frameworks/python.section.md:348]: 1. Unfortunately the override can be a bit complicated, since the "library" is accessed behind `python3Packages`. But it happens in both scenarios, so nothing was lost.
2. I am sprint-removing the ones in top-level.

That being said, I believe I will need to rewrite this whole text, showing the new vs the legacy method or something similar.

**Comment by AndersonTorres** [comment]
@DanielSidhion please review again.

File: doc/languages-frameworks/python.section.md [hunk 2]
```diff
@@ -336,21 +336,42 @@ Python versions or modules, which is why they don't go in `python3Packages`.
...
+to the attribute in `python-packages.nix`, and applying `toPythonApplication` to
+that reference. Let's take `youtube-dl` as an example:
 
 ```nix
 {
   youtube-dl = with python3Packages; toPythonApplication youtube-dl;
 }
 ```
 
+##### New Method {#topythonapplication-function-new-method}
```
Comment (DanielSidhion) [inline doc/languages-frameworks/python.section.md:363]: Feel like it's important to call this out to avoid people trying to migrate things to the new method just because they read it here.

```suggestion
New packages should avoid using the legacy method and prefer the new method.
Existing packages that already use the legacy method shouldn't be migrated to the new method to prevent changing its [override](#sec-pkg-override) interface.

##### New Method {#topythonapplication-function-new-method}
```

**Review by DanielSidhion — APPROVED** [review]
Thanks! Made a suggestion to avoid some potential confusion in the future.

File: doc/languages-frameworks/python.section.md (hunk 2 above)
Comment (AndersonTorres) [reply doc/languages-frameworks/python.section.md:366]: I added notes. Tell me what do you think.

File: doc/languages-frameworks/python.section.md [hunk 3]
```diff
@@ -336,21 +336,51 @@ Python versions or modules, which is why they don't go in `python3Packages`.
...
+that reference. Let's take `youtube-dl` as an example:
 
 ```nix
 {
   youtube-dl = with python3Packages; toPythonApplication youtube-dl;
 }
 ```
 
+::: {.note} The legacy method is deprecated and should be avoided for new
+packages; prefer the new method instead.
```
Comment (DanielSidhion) [inline doc/languages-frameworks/python.section.md:364]: You need to terminate the `:::`  block with another `:::`. It's similar to how ` ``` ` works in markdown.
The content of the note must be in its own line for it to become a note as well.

```suggestion
::: {.note}
The legacy method is deprecated and should be avoided for new packages; prefer the new method instead.
:::
```

File: doc/languages-frameworks/python.section.md [hunk 4]
```diff
@@ -336,21 +336,51 @@ Python versions or modules, which is why they don't go in `python3Packages`.
...
+```nix
+{ python3Packages }:
+
+with python3Packages; toPythonApplication youtube-dl
+```
+
+::: {.note} It's not recommended to convert expressions using the old method to
+the new one (unless you know what you're doing), since this conversio has the
+potential to bring changes in their respective [override](#sec-pkg-override)
+interfaces.
```
Comment (DanielSidhion) [inline doc/languages-frameworks/python.section.md:382]: Also fixed a typo in the text on this one.

```suggestion
::: {.note}
It's not recommended to convert expressions using the old method to
the new one (unless you know what you're doing), since this conversion has the
potential to bring changes in their respective [override](#sec-pkg-override)
interfaces.
:::
```

**Review by DanielSidhion — APPROVED** [review]
Made some suggestions to fix the syntax of the note blocks and fix a typo.

File: doc/languages-frameworks/python.section.md [hunk 5]
```diff
@@ -336,21 +336,55 @@ Python versions or modules, which is why they don't go in `python3Packages`.
...
+hierarchy, in a `callPackage` fashion. Taking the above `youtube-dl` as an
+example, the reference is registered at
+`pkgs/by-name/yo/youtube-dl/package.nix`:
+
+```nix
+{ python3Packages }:
+
+with python3Packages; toPythonApplication youtube-dl
+```
+
```
Comment (ShamrockLee) [inline doc/languages-frameworks/python.section.md:380]: Document the way to ping the Python version of a Python application exposed this way for override interface consistency. (Feel free to rephrase it.)

````suggestion

In case the top-level Python application requires a specific interpreter version, specify the canonical `python3Packages` to use in `all-packages.nix`, instead of hard-coding the desired `python3*Packages`, to keep the [override](#sec-pkg-override) interface consistent.

:::{.example #ex-toPythonApplication-by-name-ping}
# Ping the interpreter version of a Python application exposed in `pkgs/by-name`.
```nix
{
  youtube-dl = callPackage ../by-name/yo/youtube-dl/package.nix {
    python3Packages = python312Packages;
  };
}
```
:::
````

**Review by ShamrockLee — CHANGES_REQUESTED** [review]
Some minor suggestions here:
* We should probably avoid prone-to-irrelevance naming like "legacy method" and "new method", especially for document anchors, which are hard to change without breaking hyperlinks.
* Use example blocks to make the examples more discoverable and linkable.

File: doc/languages-frameworks/python.section.md (hunk 5 above)
Comment (DanielSidhion) [reply doc/languages-frameworks/python.section.md:380]: This is already documented in the [`pkgs/by-name` readme](https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#changing-implicit-attribute-defaults). I thought about making the same suggestion, but IMO it's better to have that information deduplicated. Perhaps add a pointer to the readme instead?

Comment (ShamrockLee) [reply doc/languages-frameworks/python.section.md:380]: How about

```

In case the top-level Python application requires a specific interpreter version, [change the implicit default value](https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#changing-implicit-attribute-defaults) of argument `python3Packages` to the desired package set, instead of hard-coding `python3*Packages` as an argument. This ensures consistent interpreter version [overriding](#sec-pkg-override) via `python3Package` argument.
```

**Comment by AndersonTorres** [comment]
> Some minor suggestions here:
> * We should probably avoid prone-to-irrelevance naming like "legacy method" and "new method",

My intention is to remove the legacy reference as soon as possible (right after converting the code), however it requires some testing.

> especially for document anchors, which are hard to change without breaking hyperlinks.

Good point.
What do you think about removing the session marks and rewording?

**Comment by ShamrockLee** [comment]
My laptop is awaiting repair, and I might be slow to response until setting up the machine I borrowed. Thank you for your patience.

> > Some minor suggestions here
> >
> > We should probably avoid prone-to-irrelevance naming like "legacy method" and "new method",
> >
> My intention is to remove the legacy reference as soon as possible (right after converting the code), however it requires some testing.

Does that mean there will be a bulk migration from the legacy reference approach to the new one?

**Comment by AndersonTorres** [comment]
> Does that mean there will be a bulk migration from the legacy reference approach to the new one?

Yes:

https://github.com/NixOS/nixpkgs/pull/295499

File: doc/languages-frameworks/python.section.md [hunk 6]
```diff
@@ -336,21 +336,69 @@ Python versions or modules, which is why they don't go in `python3Packages`.
...
+  youtube-dl = callPackage ../by-name/yo/youtube-dl/package.nix {
+    python3Packages = python312Packages;
+  };
+```
+
+::: {.note}
+It's not recommended to convert existing expressions using the old method to the
+new one (unless you know what you're doing), since this conversion has the
+potential to bring changes in their respective [override](#sec-pkg-override)
+interfaces.
```
Comment (ShamrockLee) [inline doc/languages-frameworks/python.section.md:399]: > > Does that mean there will be a bulk migration from the legacy reference approach to the new one?
> 
> Yes:
> 
> #295499
> 

Considering that we are working on the migration, it sounds more self-consistent to ask people to wait for the bulk migration instead of discouraging the migration.

```suggestion
Migration from the legacy method to the new one changes the [override](#sec-pkg-override)
interfaces of existing packages, making it backward incompatible.

Bulk migration is being worked on. Manual migration should only be performed when accompanied by/required by other changes.
```

**Comment by ShamrockLee** [comment]
The new "note about migration" looks great!

Do we need to remind people to follow the "[changing the implicit default value](https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#changing-implicit-attribute-defaults)" guide of name-based package directory when using a specific Python version?

File: doc/languages-frameworks/python.section.md [hunk 7]
```diff
@@ -336,21 +336,72 @@ Python versions or modules, which is why they don't go in `python3Packages`.
...
+example, the reference is registered at
+`pkgs/by-name/yo/youtube-dl/package.nix`:
+
+```nix
+{ python3Packages }:
+
+with python3Packages; toPythonApplication youtube-dl
+```
+
+Sometimes, the top-level Python application requires a specific Python
```
Comment (AndersonTorres) [inline doc/languages-frameworks/python.section.md:377]: @ShamrockLee 

> Do we need to remind people to follow the "[changing the implicit default value](https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#changing-implicit-attribute-defaults)" guide of name-based package directory when using a specific Python version?

I have done it here, line 377

**Review by ShamrockLee — APPROVED** [review]

**Comment by AndersonTorres** [comment]
Closing to clean up.

I will take the reverse approach: clean the code first.
